### PR TITLE
feat: Train the from-scratch logistic regression on the real banknote dataset (#1770)

### LINKED ISSUE #1770: i would like to add a ml model to distinguish fake currency notes
ml model using logisctic regression without importing it directly from the library
please assign this to me

### PR DESCRIPTION
**Pull request by Anijesh**
## Related Issue

Closes: #1770

- [x] Contributor (GSSoC)

---

## Description

Issue #1770 asks for a fake-currency model using **logistic regression written without
importing it from a library**. That class already exists in
`Fake_Currency_Detetction_with_logistic_regression/model.py` (added in #2113), so rather
than duplicate it in a new folder, this PR finishes the job inside the existing project.

### The problem

The from-scratch `LogisticRegression` was only ever demonstrated on **synthetic blobs**:

```python
X, y = make_classification(n_samples=500, n_features=2, ...)   # not currency data
```

So nothing in `model.py` actually touched banknotes, and the only path to real data was
`pipeline.ipynb`, which needs Kaggle credentials — you could not reproduce a currency result
by running anything in the folder.

### What this PR does

1. **`banknote_data.py`** — downloads and caches the UCI banknote authentication dataset
   (1,372 notes; variance, skewness, kurtosis, entropy from a wavelet transform), with a
   mirror fallback and **no credentials needed**. The stratified split and z-score scaling
   are written with NumPy, keeping the pipeline dependency-light and in the spirit of the
   from-scratch model.
2. **`model.py` now trains on that real data**, scales using training statistics only so
   nothing leaks from the test split, and reports held-out accuracy with a breakdown of false
   positives and missed forgeries.
3. **Fixes the loss calculation.** It added epsilon *inside* the log:
   `-mean(y·log(p + ε) + (1-y)·log(1-p + ε))`. Once the sigmoid saturates, `log(1 + ε)` is
   slightly **positive**, so the reported cross-entropy dipped below zero — visible as
   `Loss : -0.000000` during training. Clipping the probabilities instead keeps it correct:
   cross-entropy is never negative.
4. **15 offline unit tests** and a **`requirements.txt`**.

### Result

```text
Dataset: 1372 banknotes, 4 features (762 genuine / 610 forged)
Features: variance, skewness, kurtosis, entropy

Epoch : 1000 Loss : 0.080568
Training Complete!

Test accuracy : 97.45%  (274 held-out notes)
Genuine : 146 correct, 6 flagged as forged
Forged  : 121 caught,  1 missed
```

Only **1 forged note out of 122 was missed**, which is the error that matters here — a false
negative means a counterfeit passes as genuine.

---

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

---

## How Has This Been Tested?

**1. Ran `python model.py` end to end** — the output above is that real run, not an estimate.

**2. 15 unit tests, fully offline** (no download, no credentials):

```text
Ran 15 tests in 0.021s

OK
```

They test the maths rather than just the happy path:

* the sigmoid stays within `[0, 1]` even for extreme inputs, and a zero logit gives exactly 0.5
* the **analytic gradient is checked against a finite-difference estimate** (`atol=1e-5`)
* separable data is learned to >99% accuracy, and the loss decreases
* the loss stays finite **and never negative** — this is the regression test for the fix above
* the stratified split preserves class balance, has no train/test overlap, and is reproducible
* the scaler uses training statistics only, and a constant column does not divide by zero

**3. Nothing unwanted committed** — the cached dataset lands in `data/`, which the repo's
existing `*.csv` rule already ignores.

---

## Checklist

- [x] My code follows the guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly wherever it was hard to understand.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective and that my feature works.
- [ ] Any dependent changes have been merged and published in downstream modules. *(n/a)*

---

### Notes for reviewers

* Kept inside the existing folder rather than creating a new one, since the issue's request
  is already partly implemented there.
* `app.py` and `pipeline.ipynb` are untouched — the Streamlit dashboard and the kagglehub
  route still work exactly as before. This only adds a credential-free path to real data.
* The one change to existing logic is the loss clipping, which is a correctness fix with a
  test attached.
* Suggested labels: `type:feature`, `level:intermediate`, `gssoc:approved`.

### CODE REVIEW

**Comment by Anijesh** [comment]
@Sibam-Paul this closes #1770 🙏 — could you add `type:feature` and a level label when you get a chance?

Short version of what I found and did: the from-scratch `LogisticRegression` class the issue asks for **already exists** in this folder (added in #2113), but it was only ever demonstrated on `make_classification` synthetic blobs — nothing in `model.py` touched currency data, and the only route to real data was `pipeline.ipynb`, which needs Kaggle credentials.

So instead of adding a duplicate folder, I finished it in place:
- the model now trains on the real **UCI banknote authentication** dataset, downloaded and cached with no credentials
- **97.45%** test accuracy on 274 held-out notes, with only 1 forgery of 122 missed
- fixed a genuine bug in the loss: epsilon was added *inside* the log, so `log(p + ε)` went slightly positive once the sigmoid saturated and the reported cross-entropy dipped **below zero** (`Loss : -0.000000`). Clipping fixes it, and there is a regression test.
- 15 offline unit tests, including an analytic-vs-finite-difference gradient check, plus a `requirements.txt`

`app.py` and `pipeline.ipynb` are untouched.

**Comment by Anijesh** [comment]
Pushed a fix for the two CodeQL alerts on this PR (code scanning [520](https://github.com/Niketkumardheeryan/ML-CaPsule/security/code-scanning/520) / [521](https://github.com/Niketkumardheeryan/ML-CaPsule/security/code-scanning/521)) — the cache directory and file were created with whatever the process umask allowed. They are now created explicitly as `0755` and `0644`.

Verified after a real download:

```text
file : 0o644 data_banknote_authentication.csv
dir  : 0o755 data
data : 1372 banknotes, 4 features (762 genuine / 610 forged)
```

15 tests still pass and flake8 is clean. One-line change in each place, no behaviour change.

**Comment by Anijesh** [comment]
@Sibam-Paul could you add labels on this one when you review 🙏 — suggested: `type:feature`, `level:intermediate`, `gssoc:approved`. It has no labels at the moment, and I gather unlabelled PRs may not be counted for GSSoC. (Contributing from a fork, so I can't apply them myself.)

Summary: the from-scratch logistic regression now trains on the real UCI banknote dataset instead of synthetic blobs — 97.45% test accuracy, 1 of 122 forgeries missed — plus a fix for a loss bug that made cross-entropy print negative, and 15 offline tests. CodeQL alerts on the branch are fixed.

**Comment by Sibam-Paul** [comment]
@Anijesh fix the checks

**Comment by Anijesh** [comment]
@Sibam-Paul  bro check this
